proof/refine/RISCV64/Ipc_R.thy

```diff
@@ -3677,35 +3666,23 @@ lemma tcbReleaseEnqueue_corres:
 
     \<comment> \<open>ready_queues_relation\<close>
     apply (drule set_release_queue_new_state)
-    apply (wpsimp wp: threadSet_ready_queues_relation threadSet_tcbInReleaseQueue
-                simp: tcbQueueAppend_def setReleaseQueue_def tcbQueueEmpty_def
-            simp_del: comp_apply)
-    subgoal
-      by (force dest!: tcbQueueEnd_ksReleaseQueue last_in_set ready_or_release'D1
-                 simp: release_queue_relation_def tcbQueueEmpty_def)
+    apply (wpsimp wp: tcbQueueAppend_list_queue_relation_other threadSet_sched_pointers
+                      hoare_vcg_all_lift threadSet_inQ
+                simp: ready_queues_relation_def ready_queue_relation_def Let_def
+           | wps)+
+    apply (rule_tac x="release_queue s" in exI)
+    subgoal by (auto dest!: ready_or_release_disjoint simp: release_queue_relation_def not_queued_def)
 
-   \<comment> \<open>release_queue_relation\<close>
    apply (drule set_release_queue_new_state)
-   apply (clarsimp simp: tcbQueueAppend_def)
-   apply (wpsimp wp: threadSet_wp simp: setReleaseQueue_def)
-   apply (clarsimp simp: release_queue_relation_def list_queue_relation_def tcbQueueEmpty_def
-                         queue_end_valid_def)
-   apply (drule valid_sched_pointersD[where t=t])
-     apply (clarsimp simp: in_opt_pred opt_map_red obj_at'_def)
-    apply (clarsimp simp: in_opt_pred opt_map_red obj_at'_def)
-   apply (prop_tac "t \<noteq> last (release_queue s)")
-    apply (force dest!: last_in_set simp: not_in_release_q_def)
-   apply (drule (1) heap_ls_append[where new=t])
-    apply (clarsimp simp: not_in_release_q_def)
-   apply clarsimp
-   apply (drule obj_at'_prop)+
-   apply (rule conjI)
-    apply (clarsimp simp: opt_map_def)
-   apply (rule conjI)
-    apply (rule prev_queue_head_heap_upd)
-     apply (clarsimp simp: prev_queue_head_def opt_map_def split: if_splits)
-    apply (clarsimp dest!: hd_in_set simp: opt_map_def opt_pred_def split: option.splits)
-   apply (clarsimp simp: opt_map_def opt_pred_def split: if_splits option.splits)
+   apply (clarsimp simp: release_queue_relation_def)
+   apply (intro hoare_vcg_conj_lift_pre_fix)
+    apply ((wpsimp wp: tcbQueueAppend_list_queue_relation threadSet_sched_pointers | wps)+)[1]
+    apply (frule (1) valid_sched_pointersD[where t=t];
+           clarsimp simp: not_in_release_q_2_def in_opt_pred opt_map_red obj_at'_def)
+   apply (rule hoare_allI, rename_tac t')
+   apply (case_tac "t' = t"; clarsimp)
+    apply (wpsimp wp: threadSet_wp hoare_vcg_all_lift)
+   apply (wpsimp wp: threadSet_opt_pred_other)
```

Comment:
This looks like a dangerous thing to do in general -- is the precondition in the `wp` goal schematic here? If it is fixed, it is fine, but if it is schematic then doing `case_tac` means the second branch needs to work with the precondition the first branch instantiates. In that case we should leave a comment here that we are expecting that to work (that it's not an accident).

Reply:
The precondition is fixed here. This is one of those “lopsided” `corres` rules where we have one small thing happening in the abstract (adding or removing an item from a ready or release queue) and a bunch of stuff happening in the Haskell (performing a linked list operation, setting a head/end queue, and then setting a flag). So these rules go via `corres_from_valid`, which I found easiest to perform with fixed guards and preconditions.

Reply:
👍 Ok, in that case it's all fine.